Use API::Error instead of NSError on WebExtensionController and WebExtensionContext

ItsJamie9494 opened the pull request:
#### d348f170390f0ee57cafcfdb0a7040303992775c
<pre>
Use API::Error instead of NSError on WebExtensionController and WebExtensionContext
<a href="https://webkit.org/b/297301">https://webkit.org/b/297301</a>

Reviewed by Timothy Hatcher.

NSError is Cocoa-only. Using API::Error will allow for cross-platform support of WebExtensionController and WebExtensionContext.

* Source/WebCore/platform/network/cf/ResourceError.h:
* Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
(WebCore::createNSErrorFromResourceErrorBase):
(WebCore::ResourceError::nsError const):
(WebCore::ResourceError::cfError const):
* Source/WebKit/Shared/API/APIError.h:
(API::Error::create):
(API::Error::underlyingError const):
(API::Error::Error):
* Source/WebKit/Shared/Cocoa/WKNSError.mm:
(-[WKNSError _web_createTarget]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContext.mm:
(-[WKWebExtensionContext errors]):
(-[WKWebExtensionContext loadBackgroundContentWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionController.mm:
(-[WKWebExtensionController loadExtensionContext:error:]):
(-[WKWebExtensionController unloadExtensionContext:error:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
(WebKit::WebExtensionContext::createInjectedContentForScripts):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::icon):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(-[_WKWebExtensionContextDelegate webView:didFailNavigation:withError:]):
(WebKit::WebExtensionContext::recordError):
(WebKit::WebExtensionContext::clearError):
(WebKit::WebExtensionContext::load):
(WebKit::WebExtensionContext::unload):
(WebKit::WebExtensionContext::reload):
(WebKit::WebExtensionContext::loadBackgroundContent):
(WebKit::WebExtensionContext::loadBackgroundWebView):
(WebKit::WebExtensionContext::didFailNavigation):
(WebKit::WebExtensionContext::addInjectedContent):
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRules):
(WebKit::toAPI): Deleted.
(WebKit::WebExtensionContext::createError): Deleted.
(WebKit::WebExtensionContext::errors): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::load):
(WebKit::WebExtensionController::unload):
(WebKit::WebExtensionController::unloadAll):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
(WebKit::WebExtensionDynamicScripts::sourcePairForResource):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
(WebKit::WebExtensionMenuItem::icon const):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::toAPIError):
(WebKit::WebExtensionContext::createError):
(WebKit::WebExtensionContext::errors):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::recordErrorIfNeeded):
(WebKit::WebExtensionContext::safeToLoadBackgroundContent const):
(WebKit::WebExtensionContext::backgroundContentLoadError const):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:

Canonical link: <a href="https://commits.webkit.org/299009@main">https://commits.webkit.org/299009@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f9146dc3e036936a1e156d2d99dab268bf29db40

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117167 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36814 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27418 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123234 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69165 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119041 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37526 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45419 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88938 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/43624 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120100 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29879 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105090 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69422 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28955 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23184 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66886 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99295 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23375 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126370 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44053 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33115 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97611 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44408 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101305 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97397 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42747 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20668 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/40427 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18750 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43928 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49632 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43395 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46760 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45108 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContext.mm:
```diff
@@ -541,7 +544,12 @@ - (BOOL)hasContentModificationRules
 - (void)loadBackgroundContentWithCompletionHandler:(void (^)(NSError *error))completionHandler
 {
-    Ref { *_webExtensionContext }->loadBackgroundContent(makeBlockPtr(completionHandler));
+    Ref { *_webExtensionContext }->loadBackgroundContent([&](RefPtr<API::Error> outError) {
```
`outError` is wrong here since it is passed _in_, just `error` is fine.

```suggestion
    Ref { *_webExtensionContext }->loadBackgroundContent([&](RefPtr<API::Error> error) {
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionController.mm:
```diff
@@ -79,14 +80,34 @@ - (BOOL)loadExtensionContext:(WKWebExtensionContext *)extensionContext error:(NS
+    auto loadResult =  Ref { *_webExtensionController }->load(Ref { extensionContext._webExtensionContext }, internalOutError);
+
+    if (internalOutError)
+        *outError = wrapper(internalOutError);
```
Needs a null check of `outError`.
```suggestion
    if (outError && internalOutError)
        *outError = wrapper(internalError);
```

ItsJamie9494 replied:
Line 83 does a null check of outError already, setting it to nil. Do we still need to null check here?

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionController.mm:
```diff
@@ -79,14 +80,34 @@ - (BOOL)loadExtensionContext:(WKWebExtensionContext *)extensionContext error:(NS
+
+    RefPtr<API::Error> internalOutError;
+
+    auto loadResult =  Ref { *_webExtensionController }->load(Ref { extensionContext._webExtensionContext }, internalOutError);
```
It is passed _in_ not _out_.
```suggestion
    RefPtr<API::Error> internalError;

    auto loadResult =  Ref { *_webExtensionController }->load(Ref { extensionContext._webExtensionContext }, internalError);
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionController.mm:
```diff
@@ -79,14 +80,34 @@ - (BOOL)loadExtensionContext:(WKWebExtensionContext *)extensionContext error:(NS
+    auto unloadResult =  Ref { *_webExtensionController }->unload(Ref { extensionContext._webExtensionContext }, internalOutError);
+
+    if (internalOutError)
+        *outError = wrapper(internalOutError);
```
```suggestion
    RefPtr<API::Error> internalError;

    auto unloadResult =  Ref { *_webExtensionController }->unload(Ref { extensionContext._webExtensionContext }, internalError);

    if (outError && internalError)
        *outError = wrapper(internalError);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm:
```diff
@@ -122,7 +122,8 @@
 {
-    reload();
+    RefPtr<API::Error> outError = nullptr;
+    reload(outError);
```
`reload()` should have a default parameter value for `outError`, so no change is needed here.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -299,52 +235,38 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    m_errors.removeAllMatching([&](auto& error) {
+        if (error->errorCode() == errorCode)
+            return true;
+        return false;
```
```suggestion
        return error->errorCode() == errorCode;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -299,52 +235,38 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 }
 
-NSArray *WebExtensionContext::errors()
+bool WebExtensionContext::load(WebExtensionController& controller, String storageDirectory, RefPtr<API::Error>& outError)
```
This applies everywhere, but we should switch to `Expected<bool, RefPtr<API::Error>>` and drop `outError`. This can be a follow-up, unless you find it easier to do now. (Same would apply to `WebExtension` class.)

ItsJamie9494 replied:
I'll do Expected in this MR for Context/Controller, and do a follow-up for WebExtension

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -371,8 +366,9 @@ - (void)_didDeleteLocalStorage:(NSNotification *)notification
 void WebExtensionController::unloadAll()
 {
     auto contextsCopy = m_extensionContexts;
+    RefPtr<API::Error> error = nullptr;
```
Defaults to `nullptr`.
```suggestion
    RefPtr<API::Error> error;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
```diff
@@ -28,17 +28,85 @@
+    if (!customLocalizedDescription.isEmpty())
+        localizedDescription = customLocalizedDescription;
+
+    return API::Error::create({ "WKWebExtensionContextErrorDomain"_s, errorCode, { }, localizedDescription });
```
It is sad we loose the `underlyingError` on the result. It should be attached like before. This is the same for `WebExtension` errors too. We should fix that!

ItsJamie9494 replied:
AFAIU API::Error doesn't have a way to do that, so I'd have to add a way to store an API::Error inside of the API::Error, is that ok?

xeenon replied:
Yeah that sounds fine. It will need propagated to the `NSError` when it is created from an `API::Error`.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -318,9 +330,9 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
-    bool reload(NSError ** = nullptr);
+    bool load(WebExtensionController&, String storageDirectory, RefPtr<API::Error>&);
+    bool unload(RefPtr<API::Error>&);
+    bool reload(RefPtr<API::Error>&);
```
This is where `Expected<bool, RefPtr<API::Error>>` results would help, so error can be ignored if needed now that it is a reference.

xeenon reviewed (changes requested):
Looks good! Some fairly minor comments we should address first.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionController.mm:
```diff
@@ -79,14 +80,34 @@ - (BOOL)loadExtensionContext:(WKWebExtensionContext *)extensionContext error:(NS
+    if (outError && !loadResult.has_value()) {
+        *outError = wrapper(loadResult.error());
+        return false;
+    }
```
```suggestion
    if (!loadResult) {
        if (outError)
            *outError = wrapper(loadResult.error());
        return false;
    }
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionController.mm:
```diff
@@ -79,14 +80,34 @@ - (BOOL)loadExtensionContext:(WKWebExtensionContext *)extensionContext error:(NS
+    if (outError && !unloadResult.has_value()) {
+        *outError = wrapper(unloadResult.error());
+        return false;
+    }
```
```suggestion
    if (!unloadResult) {
        if (outError)
            *outError = wrapper(unloadResult.error());
        return false;
    }
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -475,24 +386,21 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
-    if (!controller->unload(*this, outError))
-        return false;
+    auto unloadResult = controller->unload(*this);
+    if (!unloadResult.has_value())
```
```suggestion
    if (!unloadResult)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -475,24 +386,21 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
-    if (!controller->load(*this, outError))
-        return false;
+    auto loadResult = controller->load(*this);
+    if (!loadResult.has_value())
```
```suggestion
    if (!loadResult)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -322,7 +315,8 @@ - (void)_didDeleteLocalStorage:(NSNotification *)notification
 
-    if (!extensionContext.load(*this, extensionDirectory, outError)) {
+    auto loadResult = extensionContext.load(*this, extensionDirectory);
+    if (!loadResult.has_value()) {
```
```suggestion
    if (!loadResult) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -362,8 +351,9 @@ - (void)_didDeleteLocalStorage:(NSNotification *)notification
-    if (!extensionContext.unload(outError))
-        return false;
+    auto unloadResult = extensionContext.unload();
+    if (!unloadResult.has_value())
```
```suggestion
    if (!unloadResult)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -302,14 +312,16 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
+        if (error)
+            recordError(*error);
+    }
+    void recordError(Ref<API::Error>);
```
```suggestion
    }

    void recordError(Ref<API::Error>);
```

xeenon reviewed (approved):
Thanks for doing this and adopting `Expected`!

ItsJamie9494 committed:
Use API::Error instead of NSError on WebExtensionController and WebExtensionContext
https://webkit.org/b/297301

Reviewed by Timothy Hatcher.

NSError is Cocoa-only. Using API::Error will allow for cross-platform support of WebExtensionController and WebExtensionContext.

* Source/WebCore/platform/network/cf/ResourceError.h:
* Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
(WebCore::createNSErrorFromResourceErrorBase):
(WebCore::ResourceError::nsError const):
(WebCore::ResourceError::cfError const):
* Source/WebKit/Shared/API/APIError.h:
(API::Error::create):
(API::Error::underlyingError const):
(API::Error::Error):
* Source/WebKit/Shared/Cocoa/WKNSError.mm:
(-[WKNSError _web_createTarget]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContext.mm:
(-[WKWebExtensionContext errors]):
(-[WKWebExtensionContext loadBackgroundContentWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionController.mm:
(-[WKWebExtensionController loadExtensionContext:error:]):
(-[WKWebExtensionController unloadExtensionContext:error:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
(WebKit::WebExtensionContext::createInjectedContentForScripts):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::icon):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(-[_WKWebExtensionContextDelegate webView:didFailNavigation:withError:]):
(WebKit::WebExtensionContext::recordError):
(WebKit::WebExtensionContext::clearError):
(WebKit::WebExtensionContext::load):
(WebKit::WebExtensionContext::unload):
(WebKit::WebExtensionContext::reload):
(WebKit::WebExtensionContext::loadBackgroundContent):
(WebKit::WebExtensionContext::loadBackgroundWebView):
(WebKit::WebExtensionContext::didFailNavigation):
(WebKit::WebExtensionContext::addInjectedContent):
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRules):
(WebKit::toAPI): Deleted.
(WebKit::WebExtensionContext::createError): Deleted.
(WebKit::WebExtensionContext::errors): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::load):
(WebKit::WebExtensionController::unload):
(WebKit::WebExtensionController::unloadAll):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
(WebKit::WebExtensionDynamicScripts::sourcePairForResource):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
(WebKit::WebExtensionMenuItem::icon const):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::toAPIError):
(WebKit::WebExtensionContext::createError):
(WebKit::WebExtensionContext::errors):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::recordErrorIfNeeded):
(WebKit::WebExtensionContext::safeToLoadBackgroundContent const):
(WebKit::WebExtensionContext::backgroundContentLoadError const):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:

Canonical link: https://commits.webkit.org/299009@main
